app/graph.go

```diff
@@ -0,0 +1,193 @@
+		writeAccessOp := acltypes.AccessOperation{
+			AccessType:         acltypes.AccessType_WRITE,
+			ResourceType:       accessOp.GetResourceType(),
+			IdentifierTemplate: accessOp.GetIdentifierTemplate(),
+		}
+		if writeNodeIDs, ok := dag.AccessOpsMap[writeAccessOp]; ok {
+			for _, wn := range writeNodeIDs {
+				writeNode := dag.NodeMap[wn]
+				// if accessOp exists already (and from a previous transaction), we need to define a dependency on the previous message (and make a edge between the two)
+				// if from a previous transaction, we need to create an edge
+				if writeNode.TxIndex < dagNode.TxIndex {
+					lastTxNode := dag.NodeMap[dag.TxIndexMap[writeNode.TxIndex]]
+					dag.AddEdge(lastTxNode.NodeID, dagNode.NodeID, &writeAccessOp)
+				}
+			}
+		}
+	case acltypes.AccessType_WRITE:
+		// if we need to do a write, we need read and write as dependencies
+		writeAccessOp := acltypes.AccessOperation{
+			AccessType:         acltypes.AccessType_WRITE,
+			ResourceType:       accessOp.GetResourceType(),
+			IdentifierTemplate: accessOp.GetIdentifierTemplate(),
+		}
+		if writeNodeIDs, ok := dag.AccessOpsMap[writeAccessOp]; ok {
+			for _, wn := range writeNodeIDs {
+				// if accessOp exists already (and from a previous transaction), we need to define a dependency on the previous message (and make a edge between the two)
+				writeNode := dag.NodeMap[wn]
+				// if from a previous transaction, we need to create an edge
+				if writeNode.TxIndex < dagNode.TxIndex {
+					// we need to get the last node from that tx
+					lastTxNode := dag.NodeMap[dag.TxIndexMap[writeNode.TxIndex]]
+					dag.AddEdge(lastTxNode.NodeID, dagNode.NodeID, &writeAccessOp)
+				}
+			}
+		}
+		readAccessOp := acltypes.AccessOperation{
+			AccessType:         acltypes.AccessType_READ,
+			ResourceType:       accessOp.GetResourceType(),
+			IdentifierTemplate: accessOp.GetIdentifierTemplate(),
+		}
+		if readNodeIDs, ok := dag.AccessOpsMap[readAccessOp]; ok {
+			for _, rn := range readNodeIDs {
+				readNode := dag.NodeMap[rn]
+				// if accessOp exists already (and from a previous transaction), we need to define a dependency on the previous message (and make a edge between the two)
+				// if from a previous transaction, we need to create an edge
+				if readNode.TxIndex < dagNode.TxIndex {
+					dag.AddEdge(readNode.NodeID, dagNode.NodeID, &readAccessOp)
+				}
+			}
+		}
+	default:
+		// unknown or something else, raise error
+		ctx.Logger().Error("Invalid AccessControlType Received")
+	}
+	// update access ops map with the latest node id using a specific access op
+	dag.AccessOpsMap[accessOp] = append(dag.AccessOpsMap[accessOp], dagNode.NodeID)
+}
+
+// returns completion signaling map and blocking signals map
+func (dag *Dag) BuildCompletionSignalMaps() (completionSignalingMap map[int]map[acltypes.AccessOperation][]CompletionSignal, blockingSignalsMap map[int]map[acltypes.AccessOperation][]CompletionSignal) {
```

Comment:
nit: it's not very easy to tell what `int` represents. Maybe we can define an explicit type `type DagNodeID int` for readability